### PR TITLE
Adding documentation for GPU support

### PR DESCRIPTION
**Pull request by vishh**
cc @apsinha @kubernetes/sig-node-pr-reviews

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/3018)

### CODE REVIEW

**Comment by vishh** [comment]
cc @devin-donnelly

**Comment by bgrant0607** [comment]
@vishh Why isn't this PR against the 1.6 branch?

**Comment by vishh** [comment]
@bgrant0607 updated the base branch.

**Comment by vishh** [comment]
@pwittrock can you do a technical review?

File: docs/user-guide/gpus.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+```yaml
+kind: pod
+apiVersion: v1
+spec:
+ containers:
+ - name: gpu-container-1
+   resources:
+	   limits:
+		alpha.kubernetes.io/nvidia-gpu: 2
```
Comment (pwittrock) [inline docs/user-guide/gpus.md]: nit: get rid of the red

File: docs/user-guide/gpus.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
+---
+assignees:
+- vishh
+title: GPU Support
+---
+
+Kubernetes includes **experimental** support for managing NVIDIA GPUs spread across nodes.
```
Comment (pwittrock) [inline docs/user-guide/gpus.md]: @devin-donnelly Should this be a task document?  "Scheduling GPUs"?

**Comment by pwittrock** [comment]
Maybe?  The only things I know about the feature, I would be learning from your document.  I can read it to make sure it makes sense from an engineering perspective.

File: docs/user-guide/gpus.md (hunk 1 above)
Comment (vishh) [inline docs/user-guide/gpus.md]: done

**Comment by vishh** [comment]
@pwittrock There isn't much technical content in this doc to require a deep technical review. It should make sense from a user perspective. So as someone new to this feature, if this doc is able to help you use GPUs, then my goal is met.

File: docs/user-guide/gpus.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+This page describes how users can consume GPUs and the current limitations.
+
+## Pre-requisites
+
+1. Kubernetes nodes have to be pre-installed with Nvidia drivers. Kubelet will not detect Nvidia GPUs otherwise.
```
Comment (pwittrock) [inline docs/user-guide/gpus.md]: How can a user check that the Kubelet will recognize the drivers?  Can they run a bash command that will tell them if the drivers are correctly installed?

File: docs/user-guide/gpus.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+2. A special **alpha** feature gate `Accelerators` has to be set to true across the system: `--feature-gates="Accelerators=true"`.
```
Comment (pwittrock) [inline docs/user-guide/gpus.md]: Can this be done on GKE?  Is there a link to a doc explaining how to set these flags in a running or new cluster?  kubeadm?  kops?  kube-up?

File: docs/user-guide/gpus.md (hunk 3 above)
Comment (vishh) [inline docs/user-guide/gpus.md]: The drivers have to be installed for kubelet to detect GPUs. If kubelet does not include GPUs in capacity, then the drivers have to be (re)installed.

File: docs/user-guide/gpus.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+```yaml
+apiVersion: v1
+kind: pod
+spec: 
+  containers: 
+    - 
+      name: gpu-container-1
+      resources: 
+        limits: 
+          alpha.kubernetes.io/nvidia-gpu: 2
```
Comment (pwittrock) [inline docs/user-guide/gpus.md]: What does the number mean?

Comment (pwittrock) [inline docs/user-guide/gpus.md]: How does this expose the GPU in a Pod?  Give a command with kubectl exec that will show the GPU is accessible by the container.

File: docs/user-guide/gpus.md (hunk 4 above)
Comment (vishh) [inline docs/user-guide/gpus.md]: I looked for a doc on feature gates and there isn't one :disappointed: 
Each k8s provider chooses to expose flags in their own manner. For GKE, an "alpha" cluster would turn on this feature automatically. I plan on documenting this in the GKE side. 
I don't know what to do with other providers.
We need a doc on "feature gates" in general.

File: docs/user-guide/gpus.md (hunk 5 above)
Comment (vishh) [inline docs/user-guide/gpus.md]: `resource-name: quantity` is the convention in our API. E.x. `cpu: 2`

File: docs/user-guide/gpus.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+- GPUs can be specified in the `limits` section only.
+- Containers (and pods) do not share GPUs.
+- Each container can request one or more GPUs.
+- It is not possible to request a portion of a GPU.
+- Nodes are expected to be homogenous, i.e. run the same GPU hardware.
+
+If your nodes are running different versions of GPUs, then use Node Labels and Node Selectors to schedule pods to appropriate GPUs.
+Following is an illustration of this workflow:
+
+As part of your Node bootstrapping, identify the GPU hardware type on your nodes and expose it as a node label.
```
Comment (pwittrock) [inline docs/user-guide/gpus.md]: Link to how is this done with kops / kubeadm?

File: docs/user-guide/gpus.md (hunk 5 above)
Comment (pwittrock) [inline docs/user-guide/gpus.md]: I meant update the doc :P

File: docs/user-guide/gpus.md (hunk 3 above)
Comment (pwittrock) [inline docs/user-guide/gpus.md]: Assume I have a machine and I don't know if the drivers are installed.  How do I check?  (Update the doc)

File: docs/user-guide/gpus.md (hunk 5 above)
Comment (vishh) [inline docs/user-guide/gpus.md]: lol. Done

File: docs/user-guide/gpus.md (hunk 6 above)
Comment (vishh) [inline docs/user-guide/gpus.md]: The script I posted should ideally work with kubeadm/kops too. I haven't tested it yet. 

@mikedanese @justinsb Is `/etc/default/kubelet` a valid configuration medium in kubeadm and kops?

**Comment by vishh** [comment]
@pwittrock ptal

**Comment by pwittrock** [comment]
Technically this looks good.  Will defer to Devin for where it belongs.

File: docs/user-guide/gpus.md (hunk 2 above)
Comment (chenopis) [inline docs/user-guide/gpus.md]: I agree w/ this. I think this should be a task titled "Scheduling GPUs".

**Comment by chenopis** [comment]
@devin-donnelly is having me find a place for it. I agree w/ @pwittrock, I think this should be a Task titled "Scheduling GPUs".

**Comment by vishh** [comment]
So move from `docs/user-guides/gpus.md` to `docs/tasks/managing-gpus/gpus.md` ?

**Comment by vishh** [comment]
@chenopis would you mind posting a patch to this PR to handle the reorganization? It might save time since this PR should have been merged as part of the v1.6 release.

**Comment by chenopis** [comment]
Looks like we will just merge this PR and then @steveperry-53 or I will open a separate PR to move it into it's proper place.

**Comment by chenopis** [comment]
This is getting moved in PR #3212

**Pull request merged**